TransientDB for storing private read write set

cendhu opened the pull request:
Need to add test and proper comments.

<!-- Provide a general summary of your changes in the Title above -->

## Description
<!-- Describe your changes in detail. -->

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Fixes #

## How Has This Been Tested?
<!-- If this PR does not contain a new test case, explain why. -->
<!-- Describe in detail how you tested your changes. -->

## Checklist:
<!-- To check a box, and an 'x': [x] -->
<!-- To uncheck box, add a space: [ ] -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff).
- [] I have either added documentation to cover my changes or this change requires no new documentation.
- [] I have either added unit tests to cover my changes or this change requires no new tests.
- [] I have run [golint](https://github.com/golang/lint) and have fixed valid warnings in code I have added or modified. This tool generates false positives so you may choose to ignore some warnings. The goal is clean, consistent, and readable code.

<!-- The continuous integration build process will run [make checks](https://github.com/hyperledger/fabric/blob/master/Makefile#L22) to confirm that tests pass and that code quality meets minimum standards. You may optionally run this locally as PRs will not be accepted until they pass. -->

Signed-off-by:

cendhu committed:
TransientDB for storing private read write set

Need to add test and proper comments.

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store_helper.go:
```diff
@@ -0,0 +1,110 @@
+type count uint8
+
+const (
+	NUM_ATTRIBUTES_IN_PRWSET_KEY      count = 4 //namespace~txId~endorserId~endorsementBlkHt
```
Instead of namespace this should be the ledgerid

manish-sethi replied:
golang convention for const - camel-case. I missed where are these tow being used...

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store_helper.go:
```diff
@@ -0,0 +1,110 @@
+
+const (
+	NUM_ATTRIBUTES_IN_PRWSET_KEY      count = 4 //namespace~txId~endorserId~endorsementBlkHt
+	NUM_ATTRIBUTES_IN_PURGE_INDEX_KEY count = 4 //namespace~endorsementBlkHt~txId~endorserId
```
Better to call it prefix, Instead of 'namespace'. 'namespace' may be confused with the chaincode namespace.

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store_helper.go:
```diff
@@ -0,0 +1,110 @@
+
+// CreateCompositeKeyForPRWSet creates a key for storing private read-write set
+// in the transient store. The structure of the key is namespace~transactionId~endorserId.
+func CreateCompositeKeyForPRWSet(ns string, txId string, endorserId string, endorsementBlkHt uint64) []byte {
```
Do we need these functions public? Better to hide the implementation detail within package

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store_helper.go:
```diff
@@ -0,0 +1,110 @@
+	partialCompositeKey = append(partialCompositeKey, []byte(ns)...)
+	partialCompositeKey = append(partialCompositeKey, compositeKeySep...)
+	partialCompositeKey = append(partialCompositeKey, []byte(txId)...)
+
```
add 'compositeKeySep' for the start key for precise prefix... (to avoid the mistake if the txid happens to be a substring of some other txid... Should not happen in the current txid impl but better to be more restrictive)

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,154 @@ limitations under the License.
+)
+
+const (
+	PRWSET_NS     string = "P" // namespace for storing private read-write set in transient store.
```
Golang convention does not permit 'PRWSET_NS'.. it should be camel-case. Please run go fmt etc.. tools once to catch all these.

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,154 @@ limitations under the License.
+}
+
+// TransientStoreProvider provides an instance of a TransientStore
 type TransientStoreProvider interface {
```
Just thinking again do we need multiple logical dbs for transient store. Or one is good enough. If we use only one logical db, the ledgerid needs to be explicitly added to the composite key. We can leave this as is for now... Lets discuss the pro/cons and can change later in a separate PR

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store_helper.go:
```diff
@@ -0,0 +1,110 @@
+	return txId, endorserId, endorsementBlkHt
+}
+
+func CreatePartialCompositeKeyForPRWSet(ns string, txId string, endKey bool) []byte {
```
better to have two separate functions with clear names 'createTxidRangeStartKey' and 'createTxidRangeEndKey'

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,154 @@ limitations under the License.
+// TransientStoreProvider provides an instance of a TransientStore
 type TransientStoreProvider interface {
-	OpenStore(namespace string) (TransientStore, error)
+	OpenStore(dbName string) (TransientStore, error)
```
'ledgerid' should be a clearer name instead of 'dbName'

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,154 @@ limitations under the License.
+	if err := store.db.Put(compositeKey, emptyValue, true); err != nil {
+		return err
+	}
+	return nil
```
Better to use a batch instead of separately calling the put two times (efficiency and consistency)

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,154 @@ limitations under the License.
+		}
+		if err := store.db.Delete(dbKey, true); err != nil {
+			return err
+		}
```
Again use a batch and keep adding deletes in the batch within this for loop. Write the batch at the end of for loop

manish-sethi reviewed (commented):
Hi Senthil - haven't yet reviewed all the files. Can you please break this PR into two. One for the impl of Transient store only (after fixing the comments)  and another for the usage of transient store in kv ledger.

cendhu commented:
I will make these changes.

cendhu committed:
Transient Store

This CR adds transient store for storing/retrieving/purging the
private read write set generated as a result of a transaction
simulation.

Unit test provides 100% coverage.

manish-sethi commented on core/ledger/kvledger/pvtrwstorage/transient_store.go:
```diff
@@ -16,26 +16,166 @@ limitations under the License.
+}
+
+// Persist stores the private read-write set of a transaction in the transient store
+func (store *transientStore) Persist(txId string, endorserId string,
```
There would be multiple private Write-sets for each tran - one per collection. For instance, if a tran touches 3 different collections. The private write-set for different collections needs to be propagated to the desired participants.
In order to meet the above requirement, please add 'collection' also as one of the parameters in the above API.

We would need to change all the APIs to take a collection

cendhu replied:
In the privateSimulationResults []byte, don’t we store `NsPvtReadWriteSet` defined in ledger/rwset.proto? Whoever needs the private read write set should have access to all collections. Hence, why do we need to differentiate private rwset in transient store using collections?

Validator will fetch all the private rwset from transient store, deserialize, and get collection-name + private read write set. 

Endorser needs to forward all collection's rwset to others. 

Won't storing `NsPvtReadWriteSet` alone suffice?

manish-sethi reviewed (commented):
Hi Senthil, one point which I did not thought earlier. Can you fix this now?
